Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,335 @@
+This makes the feature more flexible in case a user needs to shard on e.g.
+regions instead.
+
+## Generated configuration
+
+In order to illustrate configuration generator mechanics, the following
+examples use Helm templating syntax.
+Variables are based on the proposed configuration structure as defined above.
+Functions from the [sprig](https://masterminds.github.io/sprig/) library are
+used, as they are also present in Helm.
+
+Values generated by the operator, like `shard_index` are retrieved from an
+`.operator` object.
+
+### classic mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  shardingStrategy:
+    mode: 'Classic'    
+    additionalRelabelConfig: [] 
+    classic:
+      sourceLabel: '__address__'
+```
+
+And the relabel configuration being genereated through this template
+
+```yaml
+{{ .spec.shardingStrategy.additionalRelabelConfig | toYaml }}
+- source_labels: [ '{{ default "__address__" .spec.shardingStrategy.classic.sourceLabel }}' ] 
+  separator: ;
+  modulus: {{ .spec.shards }}
+  target_label: '__tmp_hash'
+  replacement: '$1'
+  action: 'hashmod'
+- source_labels: ['__tmp_hash']
+  separator: ;
+  regex: '{{ .operator.shardIndex }}'
+  replacement: '$1'
+  action: 'keep'
+```
+
+we would get the following output for `shard_index == 2`
+
+```yaml
+- source_labels: [ '__address__' ] 
+  separator: ;
+  modulus: 4
+  target_label: '__tmp_hash'
+  replacement: '$1'
+  action: 'hashmod'
+- source_labels: ['__tmp_hash']
+  separator: ;
+  regex: '2'
+  replacement: $1
+  action: 'keep'
+```
+
+### topology mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  shardingStrategy:
+    mode: 'Topology'    
+    additionalRelabelConfig: [] 
+    topology:
+      nodeLabel: 'topology.kubernetes.io/zone'
+      sourceLabel: '__meta_kubernetes_pod_label_topology_kubernetes_io_zone'
+      values:
+        - europe-west4-a
+        - europe-west4-b
+```
+
+And the relabel configuration being genereated through this template
+
+```yaml
+{{ $shardsPerZone := max 1 (.spec.shards | div (len .spec.shardingStrategy.topology.values)) }}
+{{ $promZoneIdx := .operator.shardIndex | mod $shardsPerZone }}
+
+{{ .spec.shardingStrategy.additionalRelabelConfig | toYaml }}
+- source_labels: [ '{{ default "__meta_kubernetes_pod_label_topology_kubernetes_io_zone" .spec.shardingStrategy.topology.sourceLabel }}' ]
+  separator: ;
+  regex: '{{ index .spec.shardingStrategy.topology $promZoneIdx }}'
+  replacement: '$1'
+  action: 'keep'
+- source_labels: [ '__address__' ] 
+  separator: ;
+  modulus: {{ $shardsPerZone }}
+  target_label: '__tmp_hash'
+  replacement: '$1'
+  action: 'hashmod'
+- source_labels: ['__tmp_hash']
+  separator: ;
+  regex: '{{ .operator.shardIndex | div $shardsPerZone }}'
+  replacement: '$1'
+  action: 'keep'
+```
+
+we would get the following output for `shard_index == 2`:
+
+```yaml
+- source_labels: [ '__meta_kubernetes_pod_label_topology_kubernetes_io_zone' ] 
```

Comment:
if I'm not wrong the `topology.kubernetes.io/zone` label is only set on [nodes](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone), not pods. I suspect that we'll need to attach node metadata to the service discovery and filter on `__meta_kubernetes_node_label_topology_kubernetes_io_zone`.

As a side note, endpointslices can carry topology information so in this case, node metadata might not be required (https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#topology).

Reply:
> I suspect that we'll need to attach node metadata to the service discovery

You're right! I totally forgot about this. In the original issue I had this addressed by having the label injected into a pod ... somehow.

How can we do this through prometheus?
Is this even possible?

Reply:
Ok, seems like this is going to get complicated.
According to the docs, there is no label available we could use here:

https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config

Or am I mistaken?

Reply:
> I suspect that we'll need to attach node metadata to the service discovery

This means a code change to the `kubernetes_sd` ?

Reply:
@simonpasquier If this requires a change to the `kubernetes_sd` this needs to be mentioned in the doc, right?

Reply:
I added a new section to the document to discuss this.